### ISSUE TITLE
Support authenticating to SFTP repositories via private key

### ISSUE THREAD

**Issue by mark-vieira** [gradle/gradle#1263]
### Expected Behavior

Users should be allowed to specify private key authentication when configuring `sftp://` repositories.

### Current Behavior

Currently, only basic username/password authentication is supported.

### Context

Private key auth is basically the defacto way to authenticate to remote servers at this point. This is becoming more and more the case with HTTP, which we do support, albeit in a non-ideal way via the standard Java system properties. When using SSH-based repositories however, via the `sftp://` protocol, no such method is available and only username/password authentication is supported.

**Comment by damnhandy** [comment]
Is this a duplication of #1071 ? Furthermore, did the PR associated with this issue make it into a release yet?

**Comment by dmodoomsirius** [comment]
Has this made it into release yet? I would really like this feature.

**Comment by ebourg** [comment]
If this feature is implemented it would be nice to use the keys defined in `~/.ssh/config`

**Comment by dmodoomsirius** [comment]
I would like to know if this is either possible or still on the list to be implemented.

**Comment by kaarelnu** [comment]
I too am interested in this feature.

**Comment by spricetracom** [comment]
Agreed I am more than interested in this feature, as security has become such a large focus, the lack of this feature is becoming a show stopping flaw.

**Comment by ov7a** [comment]
Sorry for the late reply.

This feature request is in the backlog of the relevant team and is prioritized by them.

**Comment by ov7a** [comment]
This issue would require a new `Credentials` implementation. For now, It's unclear whether it should have eager or lazy properties. 

See:
* https://github.com/gradle/gradle/issues/30823

Another caveat is related to implementation details. Currently, a shared instance of `Jsch` is used in shared `SftpClientFactory`. A password is set per session. However, by default `Identity` is set per `Jsch` instance, so a custom Identity repo should be provided per session to isolate the keys. 

There should also be an option for providing a [SSH config](https://github.com/mwiede/jsch/wiki/Jsch-Configuration), ideally with a helper method for the default location. It cannot be customized per session, though.
